File: app/src/main/java/com/timursoft/imtranslator/entity/SerializableConverter.java
package com.timursoft.imtranslator.entity;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import io.requery.Converter;

public abstract class SerializableConverter<T extends Serializable> implements Converter<T, byte[]> {

    private final Class<T> type;

    public SerializableConverter(Class<T> type) {
        if (type == null) {
            throw new IllegalArgumentException();
        }
        this.type = type;
    }

    @Override
    public Class<T> getMappedType() {
        return type;
    }

    @Override
    public Class<byte[]> getPersistedType() {
        return byte[].class;
    }

    @Override
    public Integer getPersistedSize() {
        return null;
    }

    @Override
    public byte[] convertToPersisted(T value) {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        ObjectOutput out = null;
        try {
            out = new ObjectOutputStream(bos);
            out.writeObject(value);
            return bos.toByteArray();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (out != null) {
                    out.close();
                }
            } catch (IOException ex) {
                // ignore close exception
            }
            try {
                bos.close();
            } catch (IOException ex) {
                // ignore close exception
            }
        }
        return new byte[0];
    }

    @Override
    public T convertToMapped(Class<? extends T> type, byte[] value) {
        ByteArrayInputStream bis = new ByteArrayInputStream(value);
        ObjectInput in = null;
        try {
            in = new ObjectInputStream(bis);
            return type.cast(in.readObject());
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                bis.close();
            } catch (IOException ex) {
                // ignore close exception
            }
            try {
                if (in != null) {
                    in.close();
                }
            } catch (IOException ex) {
                // ignore close exception
            }
        }
        return null;
    }
}

File: app/src/main/java/com/timursoft/imtranslator/entity/SerializableConverter.kt
package com.timursoft.imtranslator.entity

import io.requery.Converter
import java.io.*

abstract class SerializableConverter<T : Serializable>(private val type: Class<T>) : Converter<T, ByteArray> {

    override fun getMappedType(): Class<T> {
        return type
    }

    override fun getPersistedType(): Class<ByteArray> {
        return ByteArray::class.java
    }

    override fun getPersistedSize(): Int? {
        return null
    }

    override fun convertToPersisted(value: T): ByteArray {
        return ByteArrayOutputStream().use { bos ->
            ObjectOutputStream(bos).use { out ->
                out.writeObject(value)
            }
            bos
        }.toByteArray()
    }

    override fun convertToMapped(type: Class<out T>, value: ByteArray?): T? {
        return ByteArrayInputStream(value).use { bis ->
            ObjectInputStream(bis).use { ois ->
                type.cast(ois.readObject())
            }
        }
    }
}